src/components/containers/Panel.js

```diff
@@ -1,51 +1,139 @@
-import React, {Component} from 'react';
 import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import PanelHeader from './PanelHeader';
 import {bem} from 'lib';
-import {ChartLineIcon} from 'plotly-icons';
-class Panel extends Component {
-  render() {
-    const {visible, children} = this.props;
 
-    if (visible) {
-      return <div className={bem('panel')}>{children}</div>;
-    }
-    return null;
+export default class Panel extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      nbOfFolds: 0,
+      individualFoldStates: [],
+    };
+    this.toggleFolds = this.toggleFolds.bind(this);
+    this.toggleFold = this.toggleFold.bind(this);
   }
-}
 
-export class PanelHeader extends Component {
-  render() {
-    const {children, action} = this.props;
+  getChildContext() {
+    return {
+      individualFoldStates: this.state.individualFoldStates,
+      toggleFold: this.toggleFold,
+    };
+  }
 
+  toggleFolds() {
+    const {individualFoldStates} = this.state;
+    const hasOpen =
+      individualFoldStates.length > 0 &&
+      individualFoldStates.some(s => s !== true);
+    this.setState({
+      individualFoldStates: individualFoldStates.map(() => {
+        if (hasOpen) {
+          return true;
+        }
+        return false;
+      }),
+    });
+  }
+
+  toggleFold(index) {
+    const copy = [...this.state.individualFoldStates];
```

Comment:
this would be a good candidate for https://github.com/kolodny/immutability-helper

Reply:
I'll add this one, I think it'll be helpful, thanks!

Reply:
Maybe would be worth doing that in a single pass where we refactor other places where we have used various spread/copy tricks, rather than adding this dependency just for this case. I just wanted to advocate for this library because I've found it so, so useful in many little places :)

Reply:
Especially useful with Redux, as Ben used it in the batch repo

Reply:
hmm, well, for now, I'm just going to add it in this one place, or at least, this pr is only going to change it in this one place :) I'll open an issue for changing it throughout the editor.

Yeah, I've thought I've seen Ben use something to help with this, but I didn't find anything in the repo, it was in the batch one! thanks!

Reply:
OK, sounds good!